node/core/approval-voting/src/import.rs

```diff
@@ -380,6 +403,24 @@ async fn imported_block_info(
 
 	let babe_epoch = {
 		let (s_tx, s_rx) = oneshot::channel();
+
+		// It's not obvious whether to use the hash or the parent hash for this, intuitively. We
+		// want to use the block hash itself, and here's why:
+		//
+		// First off, 'epoch' in BABE means 'session' in other places. 'epoch' is the terminology from
+		// the paper, which we fulfill using 'session's, which are a Substrate consensus concept.
+		//
+		// In BABE, the on-chain and off-chain view of the current epoch can differ at epoch boundaries
+		// because epochs change precisely at a slot. When a block triggers a new epoch, the state of
+		// its parent will still have the old epoch. Conversely, we have the invariant that every
+		// block in BABE has the epoch _it was authored in_ within its post-state. So we use the
+		// block, and not its parent.
+		//
+		// It's worth nothing that Polkadot session changes, at least for the purposes of parachains,
+		// would function the same way, except for the fact that they're always delayed by one block.
+		// This gives us the opposite invariant for sessions - the parent block's post-state gives
+		// us the canonical information about the session index for any of its children, regardless
+		// of which slot number they might be produced at.
```

Comment:
Does this second concern pop up anywhere yet?  It will eventually since we expose babe state to parachains of course, but maybe not in polkadot code, not sure..

Reply:
This is the only place we actually have to dig into BABE in the code, so these concerns haven't appeared anywhere else.